api/src/main/java/org/eclipse/microprofile/config/ConfigProvider.java

```diff
@@ -86,7 +84,7 @@ private ConfigProvider() {
      * @return the config object for the thread context classloader
      */
     public static Config getConfig() {
-        return INSTANCE.getConfig();
+        return ConfigProviderResolver.instance().getConfig();
```

Comment:
As explained in https://github.com/eclipse/microprofile-config/issues/265#issuecomment-348287580, this should be changed to call `ConfigProviderResolver.instance().getConfig();` only once because `Config` is expected to be a singleton and has to be released later. 

Here's an example of the code:
```
        if (INSTANCE == null) {
            synchronized (RestClientBuilderResolver.class) {
                if (INSTANCE != null) {
                    return INSTANCE;
                }
                INSTANCE = ConfigProviderResolver.instance();
            }
        }

        return INSTANCE;
```

Reply:
Or the same can be done with simpler code by moving the INSTANCE field and initialization of `private static final ConfigProviderResolver INSTANCE = ConfigProviderResolver.instance();` into a separate class which would be used only from `getConfig()` methods, like this:

ConfigProvider.java:
```
public static Config getConfig() {
         return ConfigSingletonHolder.INSTANCE.getConfig();
```
new ConfigSingletonHolder.java:
```
class ConfigSingletonHolder {
   static final ConfigProviderResolver INSTANCE = ConfigProviderResolver.instance();
}
```

This subtle refactoring would mean that the initialization of INSTANCE happens when ConfigSingletonHolder class is loaded and that happens when `getConfig` method is called and no sooner. But few people understand how it works so it's probably better to use the explicit code with locks and null checks which is much more straightforward. Unfortunately much harder to get it right and thread-safe.

Reply:
the classloader behavior you mention is probably why in the test I had to add forced loading the class.  I suspect the CDI runtime in the running application causes the class to be loaded, resulting in the field being set before it was meant to (I believe the user in my case was setting it via a CDI extension, but need to confirm).